### PR TITLE
Device/event framework optimisation

### PR DESCRIPTION
**Pull request by pkendall64**
Two parts to this PR.
1. Devices that are not supported by the hardware can return `false` from their `init` function and the device framework will not call their `start`, `timeout` or `event` functions during the main loop.
2. Event publish/subscribe
   - Events fired from the code now pass an event type so only subscribers to that event type will be notified, reducing the number of calls in the main loop.
   - Device event framework extended to allow device handlers to subscribe to certain events and their `event` function will only be called if that type of event is fired.

### CODE REVIEW

File: src/lib/AnalogVbat/devAnalogVbat.cpp [hunk 1]
```diff
@@ -113,8 +114,9 @@ static int timeout()
 }
 
 device_t AnalogVbat_device = {
-    .initialize = nullptr,
+    .initialize = initialize,
     .start = start,
     .event = nullptr,
     .timeout = timeout,
+    .subscribe = 0
```
Comment (JyeSmith) [inline src/lib/AnalogVbat/devAnalogVbat.cpp:121]: Could this also be a enum, EVENT_NONE?

File: src/lib/BUTTON/devButton.cpp [hunk 2]
```diff
@@ -117,8 +113,8 @@ static int timeout()
...
 
 device_t Button_device = {
-    .initialize = nullptr,
+    .initialize = initialize,
     .start = start,
     .event = event,
-    .timeout = timeout
-};
+    .timeout = timeout,
+    .subscribe = EVENT_ARM_FLAG_CHANGED | EVENT_CONNECTION_CHANGED};
```
Comment (JyeSmith) [inline src/lib/BUTTON/devButton.cpp:120]: } on new line.

File: src/lib/DEVICE/device.h [hunk 3]
```diff
@@ -7,11 +7,32 @@
 #define DURATION_NEVER -1       // timeout() will not be called, only event()
 #define DURATION_IMMEDIATELY 0  // timeout() will be called each loop
 
+enum deviceEvent_t {
```
Comment (JyeSmith) [inline src/lib/DEVICE/device.h:10]: `EVENT_NONE = 0,` added here?

File: src/lib/LED/devLED.cpp [hunk 4]
```diff
@@ -213,7 +223,7 @@ static int event()
 
 device_t LED_device = {
     .initialize = initialize,
-    .start = event,
+    .start = start,
     .event = event,
-    .timeout = timeout
-};
+    .timeout = timeout,
+    .subscribe = EVENT_CONNECTION_CHANGED | EVENT_ENTER_BIND_MODE | EVENT_EXIT_BIND_MODE};
```
Comment (JyeSmith) [inline src/lib/LED/devLED.cpp:229]: } on new line.

File: src/lib/LED/devRGB.cpp [hunk 5]
```diff
@@ -473,5 +466,5 @@ device_t RGB_device = {
     .initialize = initialize,
     .start = start,
     .event = timeout,
-    .timeout = timeout
-};
+    .timeout = timeout,
+    .subscribe = EVENT_CONNECTION_CHANGED | EVENT_ENTER_BIND_MODE | EVENT_EXIT_BIND_MODE};
```
Comment (JyeSmith) [inline src/lib/LED/devRGB.cpp:470]: } on new line.

File: src/lib/LUA/rx_devLUA.cpp [hunk 6]
```diff
@@ -660,7 +663,8 @@ device_t LUA_device = {
   .initialize = nullptr,
   .start = start,
   .event = event,
-  .timeout = timeout
+  .timeout = timeout,
+  .subscribe = 0xFFFF // any and all events
```
Comment (JyeSmith) [inline src/lib/LUA/rx_devLUA.cpp:667]: Change to a ENUM?

File: src/lib/LUA/tx_devLUA.cpp [hunk 7]
```diff
@@ -910,10 +909,11 @@ static int start()
 }
 
 device_t LUA_device = {
-  .initialize = NULL,
+  .initialize = nullptr,
   .start = start,
   .event = event,
-  .timeout = timeout
+  .timeout = timeout,
+  .subscribe = 0xFFFF // all events
```
Comment (JyeSmith) [inline src/lib/LUA/tx_devLUA.cpp:916]: ENUM?

File: src/lib/MSPVTX/devMSPVTX.cpp [hunk 8]
```diff
@@ -389,8 +404,8 @@ static int timeout(void)
 
 device_t MSPVTx_device = {
     .initialize = initialize,
-    .start = nullptr,
+    .start = start,
     .event = event,
-    .timeout = timeout
-};
+    .timeout = timeout,
+    .subscribe = EVENT_VTX_CHANGE};
```
Comment (JyeSmith) [inline src/lib/MSPVTX/devMSPVTX.cpp:410]: } on new line.

File: src/lib/ServoOutput/devServoOutput.cpp [hunk 9]
```diff
@@ -265,4 +266,4 @@ device_t ServoOut_device = {
     .start = start,
     .event = event,
     .timeout = timeout,
-};
+    .subscribe = EVENT_CONNECTION_CHANGED};
```
Comment (JyeSmith) [inline src/lib/ServoOutput/devServoOutput.cpp:269]: } on new line.

File: src/lib/WIFI/devWIFI.cpp [hunk 10]
```diff
@@ -1335,5 +1336,5 @@ device_t WIFI_device = {
   .initialize = initialize,
   .start = start,
   .event = event,
-  .timeout = timeout
-};
+  .timeout = timeout,
+  .subscribe = EVENT_CONNECTION_CHANGED};
```
Comment (JyeSmith) [inline src/lib/WIFI/devWIFI.cpp:1340]: } on new line.

File: src/src/rx-serial/devSerialIO.cpp [hunk 11]
```diff
@@ -277,16 +277,16 @@ device_t Serial0_device = {
...
 
 #if defined(PLATFORM_ESP32)
 device_t Serial1_device = {
     .initialize = nullptr,
     .start = start,
     .event = event1,
-    .timeout = timeout1
-};
+    .timeout = timeout1,
+    .subscribe = EVENT_CONNECTION_CHANGED};
```
Comment (JyeSmith) [inline src/src/rx-serial/devSerialIO.cpp:289]: } on new line.

**Review by OGCJM11 — APPROVED** [review]

**Review by mha1 — APPROVED** [review]
LGTM, works with no obvious errors on rx and module, e.g. both serials ok, pwm ok, module screen ok.

**Review by JyeSmith — APPROVED** [review]

**Pull request merged**